### PR TITLE
Benchmark tests

### PR DESCRIPTION
**Pull request by jellespijker**
Parallelization, file processing and benchmarking added

I know it is all over the place at the moment, but if I'm gonna do proper small single topic PR's you guys are gonna hate me for sure ;-)

Converting from actual text lines to Instructions still not working as expected, but that will be my mission for next week

The changes made in this PR are not condensed to a single topic.

Since they also include:

 * build environment fixes
* testing changes in UT
* addition of a Benchmark tool to verify if a multi-threaded approach is indeed a speed-up it is a 2x increase in speed. I'm now able to process roughly 300.000 lines of GCode per second
* Added an option to switch between parallelization and sequenced processing
* Changed some some parallel execution strategies

See Jira ticket CURA-7696

### CODE REVIEW

File: GHermeneus/include/GHermeneus/Machine.h [hunk 1]
```diff
@@ -55,9 +56,9 @@ namespace GHermeneus
     {
     public:
 
-        Machine() = default;
+        Machine() : parallelExecution{ true } {};
```
Comment (nallath) [inline GHermeneus/include/GHermeneus/Machine.h:59]: Codestyle. Should be parallel_execution

File: GHermeneus/include/GHermeneus/Machine.h [hunk 2]
```diff
@@ -72,14 +73,28 @@ namespace GHermeneus
 
             // Extract the Commands from each line
             std::vector<std::optional<Instruction<SSV_T, T>>> extractedCmds(lines.size());
-            std::transform(std::execution::par_unseq, lines.begin(), lines.end(), extractedCmds.begin(), extractCmd);
-            extractedCmds.erase(std::remove_if(std::execution::par_unseq, extractedCmds.begin(), extractedCmds.end(), [](const auto& cmd){
+            if (parallelExecution)
```
Comment (nallath) [inline GHermeneus/include/GHermeneus/Machine.h:76]: codestyle: No single line if statements (always put {} around em)

File: GHermeneus/include/GHermeneus/Machine.h [hunk 3]
```diff
@@ -176,9 +215,11 @@ namespace GHermeneus
             return Instruction<SSV_T, T>(lineno, cmd, params);
         };
 
+        std::string rawGCode; //!< The raw GCode (needed to store files and make sure the data of \p gcode stays in scope
```
Comment (nallath) [inline GHermeneus/include/GHermeneus/Machine.h:218]: codestyle: should be raw_gcode

File: benchmark/bm_machine.cpp [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+//
+
+#include <benchmark/benchmark.h>
+
+#include <string_view>
+#include <iostream>
+
+#include "GHermeneus/GHermeneus.h"
+
+static const std::string simple_file{"../resources/simple.gcode"};
```
Comment (nallath) [inline benchmark/bm_machine.cpp:12]: Why hardcode this? Isn't it better to make this dynamically settable (eg; file to run it on as a parameter)?

File: benchmark/bm_machine.cpp [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+static const std::string big_file{"../resources/big.gcode"};
+
+size_t get_nol(const std::string& filename)
+{
+    std::ifstream gcode_file(filename);
+    std::string line;
+    size_t nol{0};
+    while (std::getline(gcode_file, line))
+        nol++;
```
Comment (nallath) [inline benchmark/bm_machine.cpp:21]:
```suggestion
    {
        nol++;   
    }
```

File: benchmark/bm_machine.cpp [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    return nol;
+}
+
+using namespace GHermeneus::Dialects::Marlin;
+
+static void BM_simple_file_parallel(benchmark::State& state)
```
Comment (nallath) [inline benchmark/bm_machine.cpp:27]: Codestyle: bmSimpleFileParallel

**Review by nallath — CHANGES_REQUESTED** [review]

File: GHermeneus/include/GHermeneus/Machine.h (hunk 2 above)
Comment (jellespijker) [reply GHermeneus/include/GHermeneus/Machine.h:76]: damn I did it again.
I will add a .clang-format so it won't happen again

File: benchmark/bm_machine.cpp (hunk 4 above)
Comment (jellespijker) [reply benchmark/bm_machine.cpp:12]: absolutely true.
But at the time of writing the unit and benchmark test aren't real CI tests yet. The're intended to be. For now I will leave it and add a ticket to the backlog that we create some proper tests.

Comment (jellespijker) [reply benchmark/bm_machine.cpp:12]: wont fix

**Pull request merged**